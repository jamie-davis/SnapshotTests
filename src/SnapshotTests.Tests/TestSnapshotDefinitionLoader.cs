using NUnit.Framework;
using SnapshotTests.Snapshots;
using TestConsoleLib;
using TestConsoleLib.Testing;

namespace SnapshotTests.Tests
{
    [TestFixture]
    public class TestSnapshotDefinitionLoader
    {
        #region Types that define tables

        [SnapshotDefinition("LocalTable")]
        public static class TableDef
        {
            [Unpredictable] 
            public static int Id { get; set; }
        }

        [SnapshotDefinition("LocalParentTable")]
        public static class ParentTableDef
        {
            [Unpredictable] 
            public static int Id { get; set; }
        }

        [SnapshotDefinition("LocalReferencingTable")]
        public static class RelatedTableDef
        {
            [Unpredictable] public static int Id { get; set; }
            
            [References("ParentTable", nameof(ParentTableDef.Id))] //implied unpredictable because the referenced property is unpredictable
            public static int ParentId1 { get; set; }
        }

        [SnapshotDefinition("PredictableTable")]
        public static class PredictableTable
        {
            [Predictable] 
            public static int Id { get; set; }
        }

        [SnapshotDefinition("SortedTable")]
        public static class SortedTable
        {
            [Predictable] 
            public static int Id { get; set; }

            [SortField]
            public static string Name { get; set; }
        }

        [SnapshotDefinition("DescSortedTable")]
        public static class DescSortedTable
        {
            [Predictable] 
            public static int Id { get; set; }

            [DescendingSortField]
            public static string Name { get; set; }
        }

        [SnapshotDefinition("CompoundSortTable")]
        public static class CompoundSortedTable
        {
            [Predictable] 
            public static int Id { get; set; }

            [SortField(1)]
            public static string Name { get; set; }

            [DescendingSortField(0)]
            public static int Priority { get; set; }
        }

        #endregion

        [Test]
        public void DefinitionsAreLoadedFromAssembly()
        {
            //Arrange
            var assembly = GetType().Assembly;

            //Act
            var definition = SnapshotDefinitionLoader.Load(assembly);

            //Assert
            var output = new Output();
            TableDefinitionReporter.Report(definition.Tables, output);
            output.Report.Verify();
        }

        [Test]
        public void DefinitionsAreLoadedFromEnclosingType()
        {
            //Act
            var definition = SnapshotDefinitionLoader.Load(GetType());

            //Assert
            var output = new Output();
            TableDefinitionReporter.Report(definition.Tables, output);
            output.Report.Verify();
        }

        [Test]
        public void DefinitionsAreFilteredWhenFromAssembly()
        {
            //Arrange
            var assembly = GetType().Assembly;

            //Act
            var definition = SnapshotDefinitionLoader.Load(assembly, t => t != typeof(Snapshots.TestDefinitions.TableDef));

            //Assert
            var output = new Output();
            TableDefinitionReporter.Report(definition.Tables, output);
            output.Report.Verify();
        }

        [Test]
        public void DefinitionsAreFilteredWhenLoadedFromEnclosingType()
        {
            //Arrange
            var assembly = GetType().Assembly;

            //Act
            var definition = SnapshotDefinitionLoader.Load(GetType(), t => t != typeof(TableDef));

            //Assert
            var output = new Output();
            TableDefinitionReporter.Report(definition.Tables, output);
            output.Report.Verify();
        }
    }
}
