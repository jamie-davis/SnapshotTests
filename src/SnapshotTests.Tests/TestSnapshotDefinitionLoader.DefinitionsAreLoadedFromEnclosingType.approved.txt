Table Definitions                                                                              
-----------------                                                                              
                      Exclude                                                                  
                      From       Include In                                                    
Table Name            Comparison Comparison Defined By Types                                   
--------------------- ---------- ---------- -------------------                                
LocalTable            False      False      TableDef                                           
                                                                                               
         Is      Is                                                    Referenced                
         Primary Compare Is            Is          Is       Referenced Property   Sort      Sort 
    Name Key     Key     Unpredictable Predictable Required Table Name Name       Direction Index
    ---- ------- ------- ------------- ----------- -------- ---------- ---------- --------- -----
    Id   False   False   True          False       False                                         
                                                                                               
                      Exclude                                                                  
                      From       Include In                                                    
Table Name            Comparison Comparison Defined By Types                                   
--------------------- ---------- ---------- -------------------                                
LocalParentTable      False      False      ParentTableDef                                     
                                                                                               
         Is      Is                                                    Referenced                
         Primary Compare Is            Is          Is       Referenced Property   Sort      Sort 
    Name Key     Key     Unpredictable Predictable Required Table Name Name       Direction Index
    ---- ------- ------- ------------- ----------- -------- ---------- ---------- --------- -----
    Id   False   False   True          False       False                                         
                                                                                               
                      Exclude                                                                  
                      From       Include In                                                    
Table Name            Comparison Comparison Defined By Types                                   
--------------------- ---------- ---------- -------------------                                
LocalReferencingTable False      False      RelatedTableDef                                    
                                                                                               
              Is      Is                                                     Referenced                
              Primary Compare Is            Is          Is       Referenced  Property   Sort      Sort 
    Name      Key     Key     Unpredictable Predictable Required Table Name  Name       Direction Index
    --------- ------- ------- ------------- ----------- -------- ----------- ---------- --------- -----
    Id        False   False   True          False       False                                          
    ParentId1 False   False   False         False       False    ParentTable Id                        
                                                                                               
                      Exclude                                                                  
                      From       Include In                                                    
Table Name            Comparison Comparison Defined By Types                                   
--------------------- ---------- ---------- -------------------                                
PredictableTable      False      False      PredictableTable                                   
                                                                                               
         Is      Is                                                    Referenced                
         Primary Compare Is            Is          Is       Referenced Property   Sort      Sort 
    Name Key     Key     Unpredictable Predictable Required Table Name Name       Direction Index
    ---- ------- ------- ------------- ----------- -------- ---------- ---------- --------- -----
    Id   False   False   False         True        False                                         
                                                                                               
                      Exclude                                                                  
                      From       Include In                                                    
Table Name            Comparison Comparison Defined By Types                                   
--------------------- ---------- ---------- -------------------                                
SortedTable           False      False      SortedTable                                        
                                                                                               
         Is      Is                                                    Referenced                
         Primary Compare Is            Is          Is       Referenced Property   Sort      Sort 
    Name Key     Key     Unpredictable Predictable Required Table Name Name       Direction Index
    ---- ------- ------- ------------- ----------- -------- ---------- ---------- --------- -----
    Id   False   False   False         True        False                                         
    Name False   False   False         False       False                          Ascending      
                                                                                               
                      Exclude                                                                  
                      From       Include In                                                    
Table Name            Comparison Comparison Defined By Types                                   
--------------------- ---------- ---------- -------------------                                
DescSortedTable       False      False      DescSortedTable                                    
                                                                                               
         Is      Is                                                    Referenced                 
         Primary Compare Is            Is          Is       Referenced Property   Sort       Sort 
    Name Key     Key     Unpredictable Predictable Required Table Name Name       Direction  Index
    ---- ------- ------- ------------- ----------- -------- ---------- ---------- ---------- -----
    Id   False   False   False         True        False                                          
    Name False   False   False         False       False                          Descending      
                                                                                               
                      Exclude                                                                  
                      From       Include In                                                    
Table Name            Comparison Comparison Defined By Types                                   
--------------------- ---------- ---------- -------------------                                
CompoundSortTable     False      False      CompoundSortedTable                                
                                                                                               
             Is      Is                                                    Referenced                 
             Primary Compare Is            Is          Is       Referenced Property   Sort       Sort 
    Name     Key     Key     Unpredictable Predictable Required Table Name Name       Direction  Index
    -------- ------- ------- ------------- ----------- -------- ---------- ---------- ---------- -----
    Id       False   False   False         True        False                                          
    Name     False   False   False         False       False                          Ascending  1    
    Priority False   False   False         False       False                          Descending 0    
                                                                                               
                                                                                               